src/binary_to_text/binary_encoder.ml

```diff
@@ -0,0 +1,713 @@
+    buf
+
+(* import: section 2 *)
+let imports_encode (func, table, mem, global) buf =
+  let fold_func import_encode_func buf acc data =
+    match
+      encode_from_indexed
+        (fun _ _ -> ())
+        import_encode_func ~is_local:false data buf
+    with
+    | None -> acc
+    | Some () -> acc + 1
+  in
+  let imp_buf = Buffer.create 16 in
+  let { Named.values = func_values; _ } = func in
+  let { Named.values = table_values; _ } = table in
+  let { Named.values = mem_values; _ } = mem in
+  let { Named.values = global_values; _ } = global in
+  let len =
+    List.fold_left
+      (fun acc func ->
+        let func = Indexed.get func in
+        match func with
+        | Runtime.Local _ -> acc
+        | Runtime.Imported { Imported.modul; name; assigned_name = _; desc = _ }
+          ->
+          add_bytes modul imp_buf;
+          add_bytes name imp_buf;
+          add_char '\x00' imp_buf;
+          write_U32_of_int acc imp_buf;
+          (* TODO ? *)
+          acc + 1 )
+      0 func_values
+  in
+  let len =
+    len + List.fold_left (fold_func table_import_encode imp_buf) 0 table_values
+  in
+  let len =
+    len + List.fold_left (fold_func memory_import_encode imp_buf) 0 mem_values
+  in
+  let len =
+    len
+    + List.fold_left (fold_func global_import_encode imp_buf) 0 global_values
+  in
+  write_U32_of_int len buf;
+  Buffer.add_buffer buf imp_buf
+
+(* function: section 3 *)
+let functions_encode func buf =
+  let idx = ref 0 in
+  vector_encode_from_namedvalues ~rev_values:true func
+    (fun v buf ->
+      let v = Indexed.get v in
+      match v with
+      | Runtime.Local _ ->
+        let typeidx = !idx in
+        incr idx;
+        Some (write_U32_of_int typeidx buf)
+      | Runtime.Imported _ -> None )
+    buf
+
+(* table: section 4 *)
+let tables_encode table buf =
+  vector_encode_from_namedvalues ~rev_values:true table
+    (encode_from_indexed table_encode (fun _ _ -> ()) ~is_local:true)
+    buf
+
+(* memory: section 5 *)
+let memories_encode memories buf =
+  vector_encode_from_namedvalues ~rev_values:true memories
+    (encode_from_indexed memory_encode (fun _ _ -> ()) ~is_local:true)
+    buf
+
+(* global: section 6 *)
+let globals_encode globals buf =
+  vector_encode_from_namedvalues ~rev_values:false globals
+    (encode_from_indexed global_encode (fun _ _ -> ()) ~is_local:true)
+    buf
+
+(* export: section 7 *)
+let exports_encode (exports : exports) buf =
+  let { global; mem; table; func } : exports = exports in
+  let exp_buf = Buffer.create 16 in
+  let len =
+    List.length global + List.length mem + List.length table + List.length func
+  in
+  List.iter
+    (fun export -> export_encode '\x03' export exp_buf)
+    (List.rev global);
+  List.iter (fun export -> export_encode '\x02' export exp_buf) (List.rev mem);
+  List.iter (fun export -> export_encode '\x01' export exp_buf) (List.rev table);
+  List.iter (fun export -> export_encode '\x00' export exp_buf) (List.rev func);
```

Comment:
`List.iter (export_encode '\x00' exp_buf) (List.rev func)` (once you changed the order), and same for all `List.iter` here